Raise when AR tests leak connections and fixes `shard_one` sqlite leak

zzak opened the pull request:
```
ARCONN=mysql2 bin/test

Connections leaked in AsyncHasManyAssociationsTest#test_async_load_has_many
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000052fa8 env_name="arunit" role=:writing>]

............................................................
Connections leaked in ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_establishing_a_connection_in_connected_to_block_uses_current_role_and_shard
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000053458 env_name="arunit" role=:writing shard=:shard_one>]

............
Connections leaked in ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_switching_connections_via_handler
[#<ActiveRecord::ConnectionAdapters::SQLite3Adapter:0x00000000054358 env_name="default_env" name="primary_shard_one" role=:writing shard=:shard_one>, #<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000054380 env_name="arunit" role=:writing>]

............................................................................................................................................................................................................................................................................................................................................................................
Connections leaked in MigrationTest#test_migrator_one_up_with_unavailable_lock_using_run
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000058638 env_name="arunit" role=:writing>]
```

I found that sqlite3 test had left a connection open, so maybe that is why we see the sqlite3 adapter errors in other adapter test runs like reported in #49373.

See also: #49584, #49373, #49382, #49403, #50579
/cc @eileencodes

I'm not sure this is the right fix to ensure the connection is closed properly (see the PR above), but I can't reproduce the log with it, and want to see how this behaves on other adapters.

zzak committed:
Raise when AR tests leak connections and fixes `shard_one` sqlite leak

```
Connections leaked in AsyncHasManyAssociationsTest#test_async_load_has_many
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000052fa8 env_name="arunit" role=:writing>]

............................................................
Connections leaked in ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_establishing_a_connection_in_connected_to_block_uses_current_role_and_shard
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000053458 env_name="arunit" role=:writing shard=:shard_one>]

............
Connections leaked in ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_switching_connections_via_handler
[#<ActiveRecord::ConnectionAdapters::SQLite3Adapter:0x00000000054358 env_name="default_env" name="primary_shard_one" role=:writing shard=:shard_one>, #<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000054380 env_name="arunit" role=:writing>]

............................................................................................................................................................................................................................................................................................................................................................................
Connections leaked in MigrationTest#test_migrator_one_up_with_unavailable_lock_using_run
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000058638 env_name="arunit" role=:writing>]
```

byroot commented on activerecord/test/cases/test_case.rb:
```diff
@@ -33,9 +33,33 @@ class TestCase < ActiveSupport::TestCase # :nodoc:
+
+      if @after_teardown_connections.size > @before_setup_connections.size
+        puts nil
+        puts "Connections leaked in #{location}"
```
By that I mean fail the test so it's noticed when the issue is introduced not later on an unrelated build.

Ruby core test suite has a bunch of leak detection iike that for threads etc.

zzak replied:
Done, yeah I wasn't sure if this was the right approach yet or not so didn't want to be over ambitious.

byroot commented:
I just ran into https://buildkite.com/rails/rails/builds/114847#0193caf2-0e9a-427d-a872-a322affd867a/1251-1262 (sqlite3 error in mysql2 build).

I wonder if it's caused by a similar sort of leak. e.g. one test always using Sqlite, and leaking the active connection to another test.

I haven't had time (more like motivation) to dig in your PR, but I think something like that would be valuable to catch leaks early.

byroot commented:
Actually, I think the culprit is highlighted by your PR:

```
Connections leaked in ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_switching_connections_via_handler
[#<ActiveRecord::ConnectionAdapters::SQLite3Adapter:0x00000000054358 env_name="default_env" name="primary_shard_one" role=:writing shard=:shard_one>, #<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000054380 env_name="arunit" role=:writing>]
```

zzak committed:
Ensure connection is thrown away after PessimisticLockingTest

```
Error:
PessimisticLockingTest#test_no_locks_no_wait:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000008790 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/locking_test.rb:826
```

zzak committed:
Ensure connection is thrown away after wait_for_async_query

```
Error:
AsyncHasOneAssociationsTest#test_async_load_has_one:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000012180 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/associations/has_one_associations_test.rb:979
```

zzak committed:
Ensure connection is thrown away after MigrationTest

```
Error:
MigrationTest#test_migrator_one_up_with_unavailable_lock:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000016bb0 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/migration_test.rb:1037
```

zzak committed:
Ensure pool is flushed after CreateOrFindByWithinTransactions tests

```
Error:
CreateOrFindByWithinTransactions#test_multiple_find_or_create_by_within_transactions:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000121960 env_name="test" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/relations_test.rb:2474
```

zzak committed:
Ensure connection is thrown away after AdapterConnectionTest

```
Error:
ActiveRecord::AdapterConnectionTest#test_#execute_is_retryable:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000124bc0 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/adapter_test.rb:835
```

zzak committed:
Ensure connection is thrown away and pool is flushed after AsynchronousQueriesTest with transactions

```
Error:
AsynchronousQueriesTest#test_async_query_from_transaction:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000129660 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/asynchronous_queries_test.rb:22
```

zzak committed:
Ensure connection is removed after ConnectionHandlersShardingDbTest
  test_establish_connection_using_3_levels_config_with_shards_and_replica

```
Error:
ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_establishing_a_connection_in_connected_to_block_uses_current_role_and_shard:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000129ab0 env_name="arunit" role=:writing shard=:shard_one>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/connection_adapters/connection_handlers_sharding_db_test.rb:18
```

zzak committed:
Try to clear more connection leaks

```
Error:
DefaultScopingWithThreadTest#test_default_scope_is_threadsafe:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x00000000123410 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/scoping/default_scoping_test.rb:726
```

Error:
MigrationTest#test_create_table_with_query:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x000000000493e0 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in `after_teardown'

bin/test test/cases/migration_test.rb:905
```

zzak committed:
Push remaining connection leak flake changes

I'm just trying to clean up what changes I have unstaged on my machine and then look into the transaction ones

zzak committed:
Try to make sure connection is reset after nested_deadlock_test

```
Error:
ActiveRecord::NestedDeadlockTest#test_deadlock_correctly_raises_Deadlocked_inside_nested_SavepointTransaction:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::Mysql2Adapter:0x0000000012b750 env_name="test" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test /rails/activerecord/test/cases/adapters/abstract_mysql_adapter/nested_deadlock_test.rb:36
```

zzak committed:
This is suspicious

```
Error:
ActiveRecordTest#test_.disconnect_all!_closes_all_connections:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::TrilogyAdapter:0x00000000002190 env_name="arunit" role=:writing>]
    test/cases/test_case.rb:58:in 'ActiveRecord::TestCase#after_teardown'
bin/rails test test/cases/active_record_test.rb:10
```

zzak committed:
Try clean_up_connection_handler and clear_all_connections

zzak committed:
Ignore connections from current adapter

zzak committed:
Revert "Ensure connection is thrown away after AdapterConnectionTest"

This reverts commit d5f762acd41a3af337df0cfaa5a7a0540f1b49b9.

zzak committed:
:nail_care: rubocop

zzak committed:
Revert 892e92ebf362afa447f02bb1de12e387dc2616fe

zzak commented:
With the following before_setup/after_teardown check:

```ruby
    def before_setup
      super
      @before_setup_connections = []
      ActiveRecord::Base.connection_handler.connection_pool_list.each do |pool|
        pool.connections.each do |conn|
          @before_setup_connections << conn
        end
      end
    end

    def after_teardown
      super
      check_connection_leaks
      @after_teardown_connections = []
      ActiveRecord::Base.connection_handler.connection_pool_list.each do |pool|
        pool.connections.each do |conn|
          @after_teardown_connections << conn
        end
      end

      leaked_conn = []
      (@after_teardown_connections - @before_setup_connections).map do |conn|
        next if conn.adapter_name == ActiveRecord::Base.lease_connection.adapter_name
        leaked_conn << conn
      end

      if leaked_conn.any?
        raise "Connections leaked\n#{leaked_conn.inspect}"
      end
    end
```

```
Error:
ActiveRecord::ConnectionAdapters::ConnectionHandlersShardingDbTest#test_switching_connections_via_handler:
RuntimeError: Connections leaked
[#<ActiveRecord::ConnectionAdapters::SQLite3Adapter:0x000000002e7778 env_name="default_env" name="primary_shard_one" role=:writing shard=:shard_one>]
    test/cases/test_case.rb:64:in `after_teardown'


bin/test test/cases/connection_adapters/connection_handlers_sharding_db_test.rb:106
```

I _think_ this is a reasonable check for "did we leave any connections behind from NOT the current adapter", and fails the sharding test which we were checking for.

All of the other "leaks" were probably due to lease_connection, which are checked back into the pool after with_connection finishes. They are not "active" and it's really hard to know which connections we _don't_ want to lie around otherwise.

After spending an embarrassing amount of time on this, my conclusion is that we should fix the `remove_connection` issue in #49584 and this PR could possibly be helpful to prevent future leaks like it in the future.

Since this issue only affects our internal Active Record tests, it's really hard to get movement on -- but I would really love to see some better leak detection if people have ideas or feedback on this one.

I will leave the other commits here for reference, but if we are OK with this method I will drop the db shard commit and squash.

zzak commented:
It's ok, I also didn't validate whether this is a good enough leak detector.

I closed this because I think it's probably better to start by looking at Eileen's PR and go from there.
